src/Monero/Address.cpp

```diff
@@ -0,0 +1,174 @@
+
+    if (string.size() != 95 && string.size() != 106){
+        std::cout<<"String size: "<<(string.size() == 106)<<std::endl;
+        return false;
+    }
+    std::string decodedAddress;
+    tools::base58::decode(string, decodedAddress);
+    std::cout<<"what";
+    return Address::isValid(parse_hex(decodedAddress));
+}
+
+Address::Address(const uint8_t networkBytes, PublicKey &spendKey, PublicKey &viewKey){
+
+    std::string address_hex = std::to_string(networkBytes) + hex(spendKey.bytes) + hex(viewKey.bytes) ;
+    Data checksum_hash = Hash::keccak256(parse_hex(address_hex));
+    address_hex += hex(checksum_hash.begin(), checksum_hash.begin()+4);
+    std::string base58_address = tools::base58::encode(address_hex.data());
+    std::vector<char> base58bytes = std::vector<char>(base58_address.begin(), base58_address.end());
+    std::copy(base58bytes.begin(), base58bytes.end(), this->bytes.begin());
+}
+
+Address::Address(const std::string &string){
+    // TODO add base58 decode check
+    // TODO add string validation
+//    if (!isValid(string)){
+//        throw std::invalid_argument("Invalid address data");
+//    }
+
+    const std::vector<char> addressBytes = std::vector<char>(string.begin(), string.end());
+    std::copy(addressBytes.begin(), addressBytes.begin()+95, this->bytes.begin());
+}
+
+Address::Address(const uint8_t networkBytes, PublicKey &spendKey, PublicKey &viewKey, const Data &paymentID){
+    std::string address_hex = std::to_string(networkBytes) + hex(spendKey.bytes) + hex(viewKey.bytes) + hex(paymentID);
+    Data checksum_hash = Hash::keccak256(parse_hex(address_hex));
+    address_hex += hex(checksum_hash.begin(), checksum_hash.begin()+4);
+    std::string base58_address = tools::base58::encode(address_hex.data());
+    std::vector<char> base58bytes = std::vector<char>(base58_address.begin(), base58_address.end());
+    std::copy(base58bytes.begin(), base58bytes.end(), this->bytes.begin());
+}
+
+
+std::string Address::string() const {
+    return std::string(bytes.begin(), bytes.end());
+}
+
+PrivateKey Address::privateViewKeyFromPrivateSpendKey(PrivateKey privateSpendKey) {
+    auto pskHash = TW::Hash::keccak256(privateSpendKey.bytes);
+    sc_reduce32(pskHash.data());
+    PrivateKey privateViewKey = PrivateKey(pskHash);
+
+    return privateViewKey;
+}
+
+PublicKey Address::publicKeyFromPrivateKey(PrivateKey privateKey) {
+    ge_p3 ge_p3_point;
+    TW::Data publicKeyData = TW::Data(32);
+
+    ge_scalarmult_base(&ge_p3_point, privateKey.bytes.data());
+    ge_p3_tobytes(publicKeyData.data(), &ge_p3_point);
+    PublicKey publicKey = PublicKey(publicKeyData, TWPublicKeyTypeED25519);
+
+    return publicKey;
+}
+
+std::string Address::getStealthAddress(PublicKey publicViewKey, PublicKey publicSpendKey){
+
+    auto r = randomScalar();
+    PrivateKey randomKey = PrivateKey(parse_hex(r));
+    PublicKey publicRandomKey = TW::Monero::Address::publicKeyFromPrivateKey(randomKey);
```

Comment:
`publicRandomKey` should be a parameter to this function. This way you can remove the hardcode `publicKey` and `privateKey` and make this function testable.

Reply:
:+1:  Good suggestion, would make testing this better. I think the random scalar _r_ is what actually should be the parameter. 

```
generateKeyDerivation(publicKey, privateKey, reinterpret_cast<char *>(keyDerivation.data()));
```

should actually be 
```
 generateKeyDerivation(publicViewKey, privateRandomKey, reinterpret_cast<char *>(keyDerivation.data()));
```

In the Stealth Address formula _K_<sup>0</sup> = _H_<sub>n</sub>(_r_ _K_<sup>v</sup><sub>B</sub>)_G_ + _K_<sup>s</sup><sub>B</sub>,  That part corresponds to _r_ _K_<sup>v</sup><sub>B</sub>, which does an Elliptic Curve Diffie Helman with a random _r_, and the recipient's public view key. 

Having _r_ be an input parameter is also useful because you need to use the corresponding public key for _r_ as part of transaction data when sending a transaction.

Reply:
Nice! Good match explanation. Let's remove [side effects](https://en.wikipedia.org/wiki/Side_effect_(computer_science)) as much as possible in our implementation, this will make the code much more predictable and testable.

Reply:
Gotcha. One thing I noticed when going through the trezor libraries was that they were almost all pure/functional. Will keep that in mind going forward.